Client/src/Service/Mixins/UserUploadService.ts

```diff
@@ -0,0 +1,170 @@
+        return responseBody;
+      }
+      console.log(response);
+
+      const message = responseBody.message;
+      if (!message){
+        throw "Unexpected error: " + response.status;
+      }
+      if (response.status != 422){
+        throw "Error type "+ responseBody.status + ": " + responseBody.message;
+      }
+      let errorLines = [];
+      errorLines.push("Validation failed:");
+      if(responseBody.errors.general && responseBody.errors.general.length){
+        errorLines.push(...responseBody.errors.general);
+      }
+      errorLines.push(Object.entries(responseBody.errors.byKey).map(p => p[0] + ": " + p[1]));
+
+      throw errorLines.join("\n<br>\n");
+
+    }).then(responseBody => {
+      const result = decoder(responseBody);
+      if (result.status === 'ok') return result.value;
+
+      let errorMessage = `Could not decode resource from ${options.path}:`;
+      if (result.context) {
+        errorMessage += '\n\n' + `  Problem at _${result.context}:`;
+      }
+      errorMessage += '\n\n' + `    Expected ${result.expected}, but got ${JSON.stringify(result.value)}.`;
+      throw errorMessage;
+    });
+}
+
+export default (base: ServiceBase) => {
+  function addDataset(newUserDataset: NewUserDataset): Promise<void> {
+    const metaBody = JSON.stringify({
+      datasetName: newUserDataset.name,
+      datasetType: newUserDataset.datasetType,
+      description: newUserDataset.description,
+      summary: newUserDataset.summary,
+      projects: newUserDataset.projects,
+      origin: "DIRECT_UPLOAD"
+    });
+
+    const fileBody = new FormData();
+    fileBody.append('file', newUserDataset.file);
+
+    return fetchDecodedJsonOrThrowMessage(
+      Decode.field("jobId", Decode.string),
+      {
+        path: '/user-datasets',
+        method: 'POST',
+        body: metaBody
+      }
+    ).then(({jobId}) => {
+      fetchWithCredentials(
+          '/user-datasets/'+jobId,
+          'POST',
+           fileBody
+      );
+    });
+  }
+  function listStatusDetails():  Promise<Array<UserDatasetUpload>> {
+    return fetchDecodedJsonOrThrowMessage(
+      Decode.arrayOf(Decode.combine(
+        Decode.field("id", Decode.string),
+        Decode.field("datasetId", Decode.optional(Decode.number)),
+        Decode.field("datasetName", Decode.string),
+        Decode.field("summary", Decode.string),
+        Decode.field("projects", Decode.arrayOf(Decode.string)),
+        Decode.field("status", Decode.string),
+        Decode.field("statusDetails", Decode.optional(Decode.combine(
+          Decode.field("errors", Decode.optional(Decode.combine(
+            Decode.field("general", Decode.arrayOf(Decode.string)),
+            Decode.field("byKey", Decode.objectOf(Decode.string))
+          ))),
+        ))),
+        Decode.field("stepPercent", Decode.optional(Decode.number)),
+        Decode.field("started", Decode.string),
+        Decode.field("finished", Decode.optional(Decode.string))
+      )),
```

Comment:
I changed it, and also extracted the functions, but it took work and i'm not super sure if it's better. The code didn't get simpler - just the opposite, I added indirection and didn't reuse anything - and now it's not in one place where I can look at it.

The more separate things I declare, the more I find myself against TypeScript. When it was all together, I was able to use the type of this big decoder together with assertion that `listStatusDetails` is of type `Promise<Array<UserDatasetUpload>>`, to check the converter, but when I extract that converter out I could only assert it's of type `any`.

Reply:
You can infer the decode type using `Decode.Unpack`:

```typescript
type UserDatasetUploadWithStatusDetails = Decode.Unpack<typeof statusDetailDecoder>;
```

This, combined with the type `UserDatasetUpload` defined in `wdk-client/Utils/Model`, you should be able to provide proper type annotations for these functions. We definitely don't want to use `any`. I would also suggest annotating the return type of the functions.

This localizes type errors, instead of getting errors in downstream code where it might not be completely obvious where the error originated.

Reply:
That's neat, I didn't know I can do this but I'll change it, thank you!